Add negative RBAC label matches

awly opened the pull request:
Normal `node_labels` matches are positive only - you must provide an
exact label or a regexp that matches all valid labels.

This is problematic when you want to grant access to "any label value
except X". You could put "X" in the `deny` section of a role, but this
prevents any other roles from granting access to "X". For example, users
may have many dynamically-named environments and one "prod", and they
want to only grant access to "prod" via workflow API.

This PR adds support for `{{regexp.not_match(...)}}` function (along with `{{regexp.match(...)}}`, for example `{{regexp.not_match("prod")}}`.

Fixes https://github.com/gravitational/teleport/issues/3454

benarent commented on docs/4.3/enterprise/ssh_rbac.md:
```diff
@@ -124,6 +124,11 @@ spec:
+      # matches and regular expressions.
+      # this is useful when another role may grant extended access and you
+      # don't want to exclude it entirely with a 'deny' rule below.
+      'environment': '+not production'
```
The space does feel little strange to me... I'm not 100% sure of the solution but I've written up some examples based on the bug.


### Proposal 1
```
role: +not PROD-*
```
### Proposal 2
```
role: +not(PROD-*)
```
### Proposal 2a
```
role: +not(`PROD-*`)
```
### Proposal 3
```
role: +not_regexp PROD-*
```
### Proposal 4
```
role: +not_regexp(PROD-*)
```
### Proposal 4a
```
role: +not_regexp(`PROD-*`)
```

### Example Group to Roles
```
    - {name: "groups", value: "okta-dev", roles: ["dev"]}
    - {name: "groups", value: "okta-dev", roles: ["+not_regexp(`PROD-*`)"]}
    - {name: "groups", value: "okta-dev", roles: ["+not(PROD-*)"]}
```

```
  'environment': '+not prod*'
```

awly replied:
@benarent up to you on the syntax.
Parentheses add complexity to the parsing code. Backticks too, plus I don't see why they are needed.
But if that's beneficial for users, implementation complexity is not an objection.

webvictim replied:
I like `+not_regexp()` over `+not` as it makes it more obvious how the field can be used and what sort of processing is going to apply. The brackets are a cleaner separator than a space IMO.

benarent replied:
As we've  already `regexp.replace(what, expression, replacement)` https://github.com/gravitational/teleport/issues/3374 

How about? `regexp.not(expression)`

awly replied:
I have a few concerns:
- we don't have `regexp.replace` in `node_labels`; `node_labels` are only globs or regexps, without the `{{foo}}`-style substitutions
- `regexp.not(expression)` is a valid regexp on its own; this change would not be strictly backwards-compatible

I'll update the PR to handle `+regexp.not(expr)`.
If you think it's more desirable to stay consistent with other role fields, we can make it `{{regexp.not(expr)}}` and just assume that no customer uses a value that looks like that. Let me know.

benarent replied:
I think we should stay as consistent as we can, I doubt any customers use a value that looks like this.

benarent commented on lib/utils/replace.go:
```diff
@@ -48,8 +48,19 @@ func ReplaceRegexp(expression string, replaceWith string, input string) (string,
 // SliceMatchesRegex checks if input matches any of the expressions. The
 // match is always evaluated as a regex either an exact match or regexp.
+//
+// If any expression starts with "+not ", the match is negated.
```
Need to update the comments to reflect the updated expression.

klizhentas commented:
@awly can you take a look at: 

https://github.com/gravitational/teleport/blob/master/lib/utils/parse/parse.go

I would like us to use the "ast" approach for all function calls. As you can spot above, there is a function "email.local" that also supports variable syntax, so it will support `{{regexp.not(}}` as well.

Ping me if you want to discuss in more detail.

awly commented:
@klizhentas WDYT about going all-in on https://golang.org/pkg/text/template/ instead?

It looks like https://github.com/gravitational/teleport/blob/master/lib/utils/parse/parse.go just reimplements a subset of that, except for input data access (`{{.external.foo}}` being `{{external.foo}}`, which we can plumb as another func: https://play.golang.org/p/YwBaMRbj3g7)

awly commented:
Chatted with @klizhentas, we can use `text/template` as long as it's backwards-compatible.
A small prototype to show how to get that working with our claims syntax: https://play.golang.org/p/vOtNwdvr9Dq

I'll flesh that out and re-open this PR.

awly commented:
After more chatting, @klizhentas convinced me to go with our [custom DSL](https://github.com/gravitational/teleport/blob/master/lib/utils/parse/parse.go).
`text/template` is flexible but its syntax is unfamiliar to non-Go developers.
A custom DSL gives us better UX at the cost of internal complexity.

Commit:
utils/parse: rewrite tests in the new RFD style

Commit:
utils/parse: support string literal expressions

No need to handle literal expressions (e.g. without "{{foo.bar}}"
substitutions) at the higher level. Something like "foo" is a valid
expression which always returns "foo" regardless of traits.

Commit:
utils/parse: implement Matcher

Matchers use a similar syntax to Expressions, but behave differently:
- Expressions get evaluated - they interpolate some values and return a
  final string.
- Matchers check whether some string matches a value

Matchers implement the same logic as utils.SliceMatchesRegex and add 2
new functions:
- {{regexp.match("foo")}} - match input against a raw regex
- {{regexp.not_match("foo")}} - same as match, but inverts the result

Commit:
Migrate services.MatchLabels to parse.Matcher

This should be backwards-compatible plus add the {{regexp.match(...)}}
and {{regexp.not_match(...)}} functions.

awly commented:
PTAL everyone, migrated this PR to use `lib/utils/parse`.
The new format supports `{{regexp.match(...)}}` and `{{regexp.not_match(...)}}`, along with the previous behavior.

webvictim commented:
LGTM, I'll leave it for others to take a look.

benarent reviewed (commented):
So based on the issue example, this would be 
```
role: {{regexp.not_match(`^critical$`)}}
```

awly commented:
@benarent almost:
```
kind: role
version: v3
metadata:
  name: dev
spec:
  allow:
    node_labels:
      'environment': '{{regexp.not_match("^prod$")}}'
```

fspmarshall commented on lib/utils/parse/parse.go:
```diff
@@ -149,20 +160,132 @@ func RoleVariable(variable string) (*Expression, error) {
+//
+// These expressions do not support variable interpolation (e.g.
+// `{{internal.logins}}`), like Expression does.
+func Match(value string) (Matcher, error) {
```
Nit: Probably preferable to name this something like `NewMatcher`.

klizhentas commented on lib/services/user.go:
```diff
@@ -491,8 +491,8 @@ func (u *UserV1) Check() error {
 	for _, login := range u.AllowedLogins {
-		_, err := parse.RoleVariable(login)
-		if err == nil {
+		e, err := parse.NewExpression(login)
```
New logic only returns error in case of grammatic errors and accepts literal values. Doesn't it mean that now `err` should return error here if it's not nil? Otherwise the following will be accepted: `{{regexp_not_match(...)` which probably should not be ignored as it's a typo and we should let users know

klizhentas commented on lib/utils/parse/parse.go:
```diff
@@ -149,20 +160,132 @@ func RoleVariable(variable string) (*Expression, error) {
+
+	re, err := regexp.Compile(raw)
+	if err != nil {
+		return nil, trace.BadParameter("failed parsing regexp %q: %v", raw, err)
```
May make sense to send a link to our documentation with examples. This will make our message actionable and reduce amount of support.

klizhentas commented on lib/utils/parse/parse.go:
```diff
@@ -176,30 +299,64 @@ func walk(node ast.Node) (*walkResult, error) {
+						return nil, trace.Wrap(err)
+					}
+					// If this is not_match, wrap the regexpMatcher to invert
+					// it.
```
Extra newline?

klizhentas reviewed (commented):
@awly Please take a look at my comments when you have a chance.

awly commented:
@klizhentas thanks for the comments, addressed in https://github.com/gravitational/teleport/pull/4430
